### PR TITLE
Wrap weighted histograms

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
This PR adds kube wrappers around the new WeightedHistograms and TimingHistograms (introduced in #109277), along the lines of the existing wrappers.

#### Which issue(s) this PR fixes:

This partially addresses #108272

#### Special notes for your reviewer:

This currently builds on #109277 , and after that merges I will rebase this onto master.
This is the middle of three steps.
The last step will be replacing the sample-and-watermark histograms with uses of the timing histograms.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/component-base/metrics/weighted_histogram.go [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+import (
+	"context"
+
+	"github.com/blang/semver/v4"
+	promext "k8s.io/component-base/metrics/prometheusextension"
+)
+
+// WeightedHistogram is our internal representation for our wrapping struct around prometheus
+// histograms. It implements both kubeCollector and ObserverMetric
+type WeightedHistogram struct {
```
Comment (wojtek-t) [inline staging/src/k8s.io/component-base/metrics/weighted_histogram.go:28]: Do we have a usecase for using WeightedHistogram in k8s codebase?
My understanding is that for now we don't have any - and if that's true, let's not expose that for now (and treat it purely as an implementation detail for TimingHistogram from k8s perspective).

[less code = less maintenance]

Comment (MikeSpreitzer) [reply staging/src/k8s.io/component-base/metrics/weighted_histogram.go]: done

**Comment by MikeSpreitzer** [comment]
The force-push to https://github.com/kubernetes/kubernetes/commit/7379d1c0112dc7c1601378c032daa089bf28edc1 is a rebase onto master now that that includes the prometheus extensions.

File: staging/src/k8s.io/component-base/metrics/timing_histogram_test.go [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+			if ok {
+				t.Error("Unexpected second metric", m2)
+			}
+
+			ms, err := registry.Gather()
+			assert.Equalf(t, test.expectedMetricCount, len(ms), "Got %v metrics, Want: %v metrics", len(ms), test.expectedMetricCount)
+			assert.Nil(t, err, "Gather failed %v", err)
+
+			for _, metric := range ms {
+				expectedHelp := helpPrefix + test.expectedHelp
```
Comment (wojtek-t) [inline staging/src/k8s.io/component-base/metrics/timing_histogram_test.go:119]: Instead of playing with setting expectedHelp here, please set the expectedHelp field in all structs appropriate to have the EXPERIMENAL: prefix.

[Same for vector test.]

**Review by wojtek-t — COMMENTED** [review]
/assign

Just one minor NIT from me - other than that it LGTM.

/approve

I will let @dgrisonnet - to make a final stamp though (or at least confirm he's fine with it).
/assign @dgrisonnet

**Comment by MikeSpreitzer** [comment]
The force-push to 32a9dbfbde8 addresses the review comments above.

**Comment by MikeSpreitzer** [comment]
The force-push to 18d9a38c0b9 adds a little overlooked cleanup.

**Comment by MikeSpreitzer** [comment]
af88f5762b0 adds a little more overlooked cleanup.

**Comment by MikeSpreitzer** [comment]
74e1f768e5b is still more cleanup

**Comment by MikeSpreitzer** [comment]
The force-push to 68d92494907 adds Reset to the methods of GaugeMetricVec, because why were we keeping it out anyway?

**Comment by MikeSpreitzer** [comment]
A brief conversation with @logicalhan turned up no enthusiasm for changing all the wrappers, so I will follow the existing pattern.  The force-push to b2c0c22e8f4 removes the remodeling.

File: staging/src/k8s.io/component-base/metrics/wrappers.go [hunk 3]
```diff
@@ -65,6 +68,70 @@ type GaugeMetric interface {
 	SetToCurrentTime()
 }
 
+// GaugeMetricVec is a collection of Gauges that differ only in label values.
+// This is really just one Metric.
+// It might be better called GaugeVecMetric, but that pattern of name is already
+// taken by the other pattern --- which is treacherous.  The treachery is that
```
Comment (wojtek-t) [inline staging/src/k8s.io/component-base/metrics/wrappers.go:74]: Taken by what?
I see "CounterVecMetric" which seems better than what you have here.

File: staging/src/k8s.io/component-base/metrics/wrappers.go [hunk 4]
```diff
@@ -65,6 +68,70 @@ type GaugeMetric interface {
...
+// WithLabelValues can return an object that is permanently broken (i.e., a noop).
+type GaugeMetricVec interface {
+	Set(value float64, labelValues ...string)
+	Inc(labelValues ...string)
+	Dec(labelValues ...string)
+	Add(delta float64, labelValues ...string)
+	SetToCurrentTime(labelValues ...string)
+
+	SetForLabels(value float64, labels map[string]string)
```
Comment (wojtek-t) [inline staging/src/k8s.io/component-base/metrics/wrappers.go:83]: Do we really need those methods? I don't see them implemented by e.g. HistogramVec, and so I don't see a reason why we really would need them here. I really prefer consistency...

**Comment by MikeSpreitzer** [comment]
@wojtek-t , thanks for your latest review.  All your comments are about the same issue, so I will address it here coherently.  Both the HistogramVec and CounterVec wrappers have the problem, since they both follow the same pattern.  One critical part of the problem is in the code that extracts a Vec member.  It starts like this:

```
func (v *CounterVec) WithLabelValues(lvs ...string) CounterMetric {
	if !v.IsCreated() {
		return noop // return no-op counter
	}
```

The Vec is not "created" until it is registered in at least one registry.  The `WithLabelValues` method is preceded by a comment that is not wrong, but did not make the important consequence blindingly obvious to me.  That consequence is that if the client holds on to the result of this call then the client will have a permanent noop.

Normally that is not a problem, because the common pattern for using such a Vec is code like the following.

```
    myVec.WithLabelValues("v1", "v2").Inc()
```

That code will do a no-op whenever it runs before `myVec` is registered, and increment an actual counter after the registration.

Also, that code has the overhead of extracting the relevant member from the Vec, and constructing an object that escapes to the heap, for every increment.

In APF we have SampleAndWatermark _client_ code that takes the opposite approach.  The SampleAndWatermark code currently uses the word "Generator" instead of "Vec", but it is essentially the same thing (and you can see that this PR switches to using "Vec" at that layer).  The SampleAndWatermark _client_ extracts the relevant members at package initialization time, which is before the relevant registration.  It works for SampleAndWatermark histograms because their _implementation_ is layered on ordinary wrapped histograms and uses members according to the pattern above. 

This PR proposes to replace SampleAndWatermark histogram "generators" with TimingRatioHistogram "vectors" and follow the usual vector pattern.  But simply making the TimingRatioHistogram implementation follow exactly the usual vector pattern breaks the client layer (the maxinflight.go and priority-and-fairness.go filters), which now latches onto permanent noops at package initialization time. So I had a few thoughts about improvements.

1. The usual pattern for using a vector member has an unnecessary heap object creation and collection.  We get more efficient code if we combine the member extraction with the member use into a single method that avoids the unnecessary heap thrashing.
2. The usual pattern for using a vector member has unnecessary computation in the cases where the call site is using constant label values (either totally constant, or constant per some other object at hand).  Better to extract the vector member up front, to avoid unnecessarily repeating the computation.
3. The existing method for extracting a vector member is treacherous --- it sometimes returns a noop without letting the client know about it.  Better to let the client know.  So I extended WithLabelValues and WithLabel with a way for the client to know whether it got shortchanged due to registration not happening yet.
4. In some cases a call site that wants to use a single Vec member and knows which one before registration is willing to pay the cost of the intermediate heap object.  It is neater to capture an object that implements the member interface at initialization time.  Even neater if this member extraction will return the more efficient post-registration member if called post registration (which it discerns by looking at the extended return from WithLabelValues).  That is what the WithLabelValuesSafe method is for.

**Comment by MikeSpreitzer** [comment]
I did a round of revisions, with an eye toward making the usual pattern available.

I removed the combined methods (e.g., `Set(float64, ...string)`) because it is not more efficient at this layer: WithLabelValues does not create an additional heap object and the rest of the member extraction costs are paid regardless.

I changed names so that the name `WithLabelValues` has the usual signature and semantics.  The additional method `WithLabelValuesChecked` exposes the extended semantics.

File: staging/src/k8s.io/component-base/metrics/timing_histogram.go [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+// will return a noop gauge and a nil error.
+// If called before this vector has been registered in
+// at least one registry, will return a noop gauge and
+// an error that passes ErrIsNotRegistered.
+// If called with a syntactic problem in the labels, will
+// return a noop gauge and an error about the labels.
+// If none of the above apply, this method will return
+// the appropriate vector member and a nil error.
+func (v *timingHistogramVec) WithLabelValuesChecked(lvs ...string) (GaugeMetric, error) {
+	if v.IsHidden() {
```
Comment (wojtek-t) [inline staging/src/k8s.io/component-base/metrics/timing_histogram.go:156]: OK - thanks for the explanation in:
https://github.com/kubernetes/kubernetes/pull/109729#issuecomment-1121353342

That makes sense to me and I'm fine with that.
But I want @dgrisonnet or @logicalhan to chime in here too.

File: staging/src/k8s.io/component-base/metrics/wrappers.go (hunk 3 above)
Comment (wojtek-t) [reply staging/src/k8s.io/component-base/metrics/wrappers.go:74]: friendly ping

File: staging/src/k8s.io/component-base/metrics/wrappers.go [hunk 6]
```diff
@@ -65,6 +68,78 @@ type GaugeMetric interface {
...
+	// The returned bool indicates whether there was a change.
+	// The return will certainly be `false` if the given label set has the wrong
+	// set of label names.
+	Delete(map[string]string) bool
+
+	// Reset removes all the members
+	Reset()
+}
+
+// PreContextGaugeMetricVec is something that can construct a GaugeMetricVec
```
Comment (wojtek-t) [inline staging/src/k8s.io/component-base/metrics/wrappers.go:123]: As suggested above - can we get rid of those interfaces? I think they add additional complexity for lack of clear value.

File: staging/src/k8s.io/component-base/metrics/timing_histogram.go [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+	*TimingHistogramOpts
+	nowFunc func() time.Time
+	lazyMetric
+	originalLabels []string
+}
+
+// NewTimingHistogramVec returns an object which satisfies kubeCollector and
+// wraps the promext.timingHistogramVec object.  Note well the way that
+// behavior depends on registration and whether this is hidden.
+func NewTimingHistogramVec(opts *TimingHistogramOpts, labels []string) PreContextAndRegisterableGaugeMetricVec {
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/component-base/metrics/timing_histogram.go]: With the alignment of WithLabelValues and of With, Gauge and TimingHistogram now implement the same interfaces, as shown in the latest revision.

[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/component-base/metrics/timing_histogram.go]: The fact that these two things implement the same interface is an important insight from @beorn7 , and I think the code should clearly say that.

File: staging/src/k8s.io/component-base/metrics/counter.go [hunk 8]
```diff
@@ -106,9 +107,14 @@ type CounterVec struct {
 	originalLabels []string
 }
 
+var _ kubeCollector = &CounterVec{}
+
+// var _ CounterVecMetric = &CounterVec{}
```
Comment (wojtek-t) [inline staging/src/k8s.io/component-base/metrics/counter.go:112]: Uncomment or remove

File: staging/src/k8s.io/component-base/metrics/wrappers.go [hunk 9]
```diff
@@ -65,6 +68,94 @@ type GaugeMetric interface {
 	SetToCurrentTime()
 }
 
+// PreContextGaugeMetric binds a Context to make a GaugeMetric
+type PreContextGaugeMetric interface {
+	// WithContext creates a GaugeMetric that uses the given Context
+	WithContext(ctx context.Context) GaugeMetric
+}
```
Comment (wojtek-t) [inline staging/src/k8s.io/component-base/metrics/wrappers.go:75]: This type isn't used anywhere else that the interface below. I'm somewhat against introducing interfaces for the sake of introducing them. Why can't we inline it below?

File: staging/src/k8s.io/component-base/metrics/wrappers.go [hunk 10]
```diff
@@ -65,6 +68,94 @@ type GaugeMetric interface {
...
+
+// PreContextAndRegisterableGaugeMetric extends GaugeMetric with registration, collection, and context binding
+type PreContextAndRegisterableCollectorGaugeMetric interface {
```
Comment (wojtek-t) [inline staging/src/k8s.io/component-base/metrics/wrappers.go:78]: > The fact that these two things implement the same interface is an important insight from [@beorn7](https://github.com/beorn7) , and I think the code should clearly say that.

I don't think it convinces me to actually buy those hard to follow interfaces.
What really matters in my opinion is the fact that TimingHistogram has exactly the same interface as GaugeMetric.
The fact of registerable, kubeCollector etc. are implementation details.

So I agree that we need conditions like:
```
var _ GaugeMetric = &TimingHistogram{}
var _ GaugetVecMetric = &TimingHistogramVec{}
```

and the need to GaugeVecMetric interface below, but I don't buy the need for any other interface here.
They introduce more congnitive burden than gain.

**Comment by MikeSpreitzer** [comment]
@wojtek-t : revised again
@dgrisonnet : can you please take a look?

File: staging/src/k8s.io/component-base/metrics/timing_histogram.go (hunk 5 above)
Comment (dgrisonnet) [reply staging/src/k8s.io/component-base/metrics/timing_histogram.go:164]: The approach taken to noop the metrics until they are initialized is correct. The deferred initialization of the metrics is normal behavior since the Kubernetes version is held by the Regsitry and because of that, it is impossible to determine if a metric is deprecated until it is registered.

As per handling whether the metrics is Hidden or not, that shouldn't be necessary since this is already taken care of when creating the underlying metric.
- some preprocessing will determine if the metric should be hidden: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/metric.go#L90-L100
- after this preprocessing, if the metrics is marked as hidden, it will not be created: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/metric.go#L146-L149

File: staging/src/k8s.io/component-base/metrics/gauge.go [hunk 11]
```diff
@@ -130,40 +140,67 @@ func (v *GaugeVec) initializeDeprecatedMetric() {
 	v.initializeMetric()
 }
 
-// Default Prometheus behavior actually results in the creation of a new metric
-// if a metric with the unique label values is not found in the underlying stored metricMap.
+func (v *GaugeVec) WithLabelValuesChecked(lvs ...string) (GaugeMetric, error) {
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/component-base/metrics/gauge.go:143]: It is not just an implementation detail, some clients want to know about the error.  #109742 has an example.

File: staging/src/k8s.io/component-base/metrics/gauge.go [hunk 12]
```diff
@@ -130,40 +140,67 @@ func (v *GaugeVec) initializeDeprecatedMetric() {
...
 // has been registered to a metrics registry.
 func (v *GaugeVec) WithLabelValues(lvs ...string) GaugeMetric {
+	ans, err := v.WithLabelValuesChecked(lvs...)
+	if err == nil || ErrIsNotRegistered(err) {
+		return ans
+	}
+	panic(err)
+}
+
+func (v *GaugeVec) WithChecked(labels map[string]string) (GaugeMetric, error) {
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/component-base/metrics/gauge.go:180]: It is not just an implementation detail, some clients want to know about the error. https://github.com/kubernetes/kubernetes/pull/109742 has an example.

File: staging/src/k8s.io/component-base/metrics/timing_histogram.go [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+	"time"
+
+	"github.com/blang/semver/v4"
+	promext "k8s.io/component-base/metrics/prometheusextension"
+)
+
+// TimingHistogram is our internal representation for our wrapping struct around timing
+// histograms. It implements both kubeCollector and GaugeMetric
+type TimingHistogram struct {
+	GaugeMetric
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/component-base/metrics/timing_histogram.go:30]: The `promext.TimingHistogram` interface has additional methods that would be exposed, by virtue of extending `Collector`.

**Comment by wojtek-t** [comment]
I think I'm roughly fine with it, but @dgrisonnet seems to have objections in some places so those needs to be resolved first (he is more familiar with this codebase than I am).

File: staging/src/k8s.io/component-base/metrics/timing_histogram.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/component-base/metrics/timing_histogram.go:30]: I just pushed a revision that splits the difference, introducing an interface that defines what aspects of the underlying TimingHistogram to promote.

File: staging/src/k8s.io/component-base/metrics/timing_histogram.go [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+		nowFunc:             nowFunc,
+		lazyMetric:          lazyMetric{},
+	}
+	h.setPrometheusHistogram(noopMetric{})
+	h.lazyInit(h, BuildFQName(opts.Namespace, opts.Subsystem, opts.Name))
+	return h
+}
+
+// setPrometheusHistogram sets the underlying KubeGauge object, i.e. the thing that does the measurement.
+func (h *TimingHistogram) setPrometheusHistogram(histogram promext.TimingHistogram) {
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply staging/src/k8s.io/component-base/metrics/timing_histogram.go:68]: Being an interface, taking a pointer is not something to do.

**Comment by MikeSpreitzer** [comment]
@dgrisonnet , I pushed a revision that responds to your review.

The revision made github stop showing the discussion around the handling of hidden vector members.  I am not sure I exactly understand your point, but thanks for catching that the hidden bit is not defined before creation.  I deliberately want to treat created and hidden differently from not created, because I think they arise in different situations and should be handled differently (again, refer to the higher level code in #109742).  The interface comments are in terms of registration rather than creation, because I think registration is the more prominent thing to clients.

File: staging/src/k8s.io/component-base/metrics/gauge.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (dgrisonnet) [reply staging/src/k8s.io/component-base/metrics/gauge.go:143]: sounds good, I was wondering why you wanted to add new methods, but it makes sense to have them

File: staging/src/k8s.io/component-base/metrics/timing_histogram.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (dgrisonnet) [reply staging/src/k8s.io/component-base/metrics/timing_histogram.go:68]: indeed, I mistook it for a struct

**Comment by MikeSpreitzer** [comment]
@dgrisonnet : thanks for the reviews!

**Comment by wojtek-t** [comment]
Just for posterity - the most recent changes LGTM.
